[Enhancement] Introduce distribution_key_column_names in TabletSchema

srlch opened the pull request:
## What I'm doing:
This is a prerequisite for tablet spliting. We introduce distribution_key_column_names in TabletSchema to ensure the ability to use them for data filtering base on hash in the future pr.

We choose physical column names instead of column unique id because column names are easier to handle compatibility issues for the tables which create in the older version without unique id in FE.

Fixes #59134 

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.5
  - [ ] 3.4
  - [ ] 3.3

srlch commented on gensrc/thrift/AgentService.thrift:
```diff
@@ -55,6 +55,7 @@ struct TTabletSchema {
     10: optional list<i32> sort_key_idxes
     11: optional list<i32> sort_key_unique_ids
     12: optional i32 schema_version;
+    13: optional list<string> distribution_key_column_names;
```
Actually i use base/physical column name here which keeps unchanged even after rename column. But it seems that cid is much more simpler. Cid for distribution keys will not affected by fast schema evolution. I will change it into cid later.

srlch replied:
Fixed

xiangguangyxg commented on gensrc/proto/tablet_schema.proto:
```diff
@@ -96,6 +96,7 @@ message TabletSchemaPB {
     repeated uint32 sort_key_unique_ids = 13;
     repeated TabletIndexPB table_indices = 14;
     optional int32 compression_level = 15 [default=-1];
+    repeated uint32 distribution_key_column_ids = 16;
```
id or unique id ?

wyb replied:
why not use `List<ColumnId.id>`?
`ColumnId.id` is not changed after table is created.

srlch replied:
Yes, this is what i have done before, save the base/physical column name `(ColumnId.id)`. But it seems that cid of distribution keys can not be changed by fast schema evolution so use cid will be more simpler than using physical col name. Any suggestion for this? I'm not 100% sure whether something i have missed if use cid here.

alvin-celerdata replied:
Just to confirm, what is the difference between ColumnId.Id and cid? And why is it simpler to use cid?

srlch replied:
`ColumnId.Id` is actually a column name but not numeric id. Since v3.3, we get two kinds of column names says `Column.getName` and `Column.getColumnId.id()` (original col name, TabletSchema actually saves `Column.getColumnId.id`). We build some mapping between them and rename columns will always change the first one but not the other one. So `Column.getColumnId.id` is unique and immutable after the table is created.

srlch replied:
Simpler may not be very accurate, using cid would be easier to understand what this field actually represents.

srlch replied:
Not 100% sure, but I guess it is more easy to implement to introduce `Column.getColumnId.id` for renaming because for BE side, nothing need to be changed.

wyb commented on be/src/storage/metadata_util.cpp:
```diff
@@ -398,6 +398,10 @@ Status convert_t_schema_to_pb_schema(const TTabletSchema& tablet_schema, uint32_
         schema->set_bf_fpp(tablet_schema.bloom_filter_fpp);
     }
+
+    for (const auto& cid : tablet_schema.distribution_key_column_ids) {
```
check `tablet_schema.__isset.distribution_key_column_ids`

srlch replied:
Fixed

srlch committed:
[Enhancement] Introduce distribution_key_column_names in TabletSchema

## What I'm doing:
This is a prerequisite for tablet spliting. We introduce distribution_key_column_names
in TabletSchema to ensure the ability to use them for data filtering base on hash in the
future pr.

We choose physical column names instead of column unique id because column names are easier to handle
compatibility issues for the tables which create in the older version without unique id in FE.

Signed-off-by: srlch <linzichao@starrocks.com>

srlch committed:
fix

Signed-off-by: srlch <linzichao@starrocks.com>

srlch committed:
fix

Signed-off-by: srlch <linzichao@starrocks.com>

srlch committed:
fix

Signed-off-by: srlch <linzichao@starrocks.com>

srlch committed:
fix

Signed-off-by: srlch <linzichao@starrocks.com>
